Unable to build on Arch Linux

Hi! I'm getting some errors while trying to build on Arch Linux...

The first error I got was this one:

```
$ cmake -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_COMPILER=/usr/bin/clang++
-- The CXX compiler identification is Clang 16.0.6
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "1.8.1") 
-- Pipewire_INCLUDE_DIR: /usr/include/pipewire-0.3
-- Pipewire_LIBRARY: /usr/lib/libpipewire-0.3.so
-- SPA_INCLUDE_DIR: /usr/include/spa-0.2
-- Components: avcodec;avdevice;avfilter;avformat;avresample;avutil;swresample;swscale
-- FFMpeg_avcodec: FOUND
-- FFMpeg_avdevice: FOUND
-- FFMpeg_avfilter: FOUND
-- FFMpeg_avformat: FOUND
CMake Error at cmake/FindFFMpeg.cmake:95 (message):
  FFMpeg: Not all required components could be found
Call Stack (most recent call first):
  CMakeLists.txt:31 (find_package)


-- Configuring incomplete, errors occurred!
```

Which I was able to "fix" by removing every reference to `avresample`, since that library [has been deprecated](https://archlinux.org/todo/libavresample-drop/) on Arch Linux for a very long time:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index e4a504e..c7b3dbc 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -35,7 +35,6 @@ find_package(
         avdevice
         avfilter
         avformat
-        avresample
         avutil
         swresample
         swscale
diff --git a/cmake/FindFFMpeg.cmake b/cmake/FindFFMpeg.cmake
index 0facb15..89914bb 100644
--- a/cmake/FindFFMpeg.cmake
+++ b/cmake/FindFFMpeg.cmake
@@ -85,7 +85,6 @@ try_find_ffmpeg_targets(
         avdevice
         avfilter
         avformat
-        avresample
         avutil
         swresample
         swscale
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 7085a3a..945c0be 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -65,7 +65,6 @@ target_link_libraries(shadow-cast-obj
     FFMpeg::avdevice
     FFMpeg::avfilter
     FFMpeg::avformat
-    FFMpeg::avresample
     FFMpeg::avutil
     FFMpeg::swresample
     FFMpeg::swscale
```

But now I got a different error, hopefully unrelated to the removal of `avresample`:

```
$ cmake --build build
[  2%] Building CXX object src/CMakeFiles/shadow-cast-obj.dir/av/buffer.cpp.o
Checking /tmp/shadow-cast/src/av/buffer.cpp ...
Checking /tmp/shadow-cast/src/av/buffer.cpp: NDEBUG=1...
Active checkers: 59/565
[  4%] Building CXX object src/CMakeFiles/shadow-cast-obj.dir/av/buffer_pool.cpp.o
Checking /tmp/shadow-cast/src/av/buffer_pool.cpp ...
Checking /tmp/shadow-cast/src/av/buffer_pool.cpp: NDEBUG=1...
Active checkers: 59/565
[  7%] Building CXX object src/CMakeFiles/shadow-cast-obj.dir/av/codec.cpp.o
Checking /tmp/shadow-cast/src/av/codec.cpp ...
Checking /tmp/shadow-cast/src/av/codec.cpp: NDEBUG=1...
Active checkers: 59/565
/tmp/shadow-cast/src/av/codec.cpp:24:30: error: no matching constructor for initialization of 'sc::BorrowedPtr<AVCodec>' (aka 'BorrowedPtr<AVCodec>')
    sc::BorrowedPtr<AVCodec> video_encoder { avcodec_find_encoder_by_name(
                             ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/shadow-cast/src/./utils/borrowed_ptr.hpp:10:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const AVCodec *' to 'const BorrowedPtr<AVCodec>' for 1st argument
struct BorrowedPtr
       ^
/tmp/shadow-cast/src/./utils/borrowed_ptr.hpp:10:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const AVCodec *' to 'BorrowedPtr<AVCodec>' for 1st argument
/tmp/shadow-cast/src/./utils/borrowed_ptr.hpp:21:5: note: candidate template ignored: constraints not satisfied [with U = const AVCodec]
    BorrowedPtr(U* p)
    ^
/tmp/shadow-cast/src/./utils/borrowed_ptr.hpp:22:14: note: because 'std::is_convertible_v<const AVCodec *, pointer>' evaluated to false
    requires(std::is_convertible_v<U*, pointer>)
             ^
/tmp/shadow-cast/src/./utils/borrowed_ptr.hpp:28:5: note: candidate template ignored: could not match 'BorrowedPtr<U>' against 'const AVCodec *'
    BorrowedPtr(BorrowedPtr<U> const& other)
    ^
/tmp/shadow-cast/src/./utils/borrowed_ptr.hpp:18:5: note: candidate constructor not viable: requires 0 arguments, but 1 was provided
    BorrowedPtr() = default;
    ^
1 error generated.
make[2]: *** [src/CMakeFiles/shadow-cast-obj.dir/build.make:104: src/CMakeFiles/shadow-cast-obj.dir/av/codec.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:102: src/CMakeFiles/shadow-cast-obj.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

Any help is appreciated. 

Thank you!

That seems to have fixed the previous error, nice. Here's a new one though:

```
In file included from /tmp/shadow-cast/src/av/sample_format.cpp:1:
/tmp/shadow-cast/src/av/sample_format.hpp:190:31: error: no type named 'uint32_t' in namespace 'std'; did you mean simply 'uint32_t'?
auto is_sample_rate_supported(std::uint32_t requested,
                              ^~~~~~~~~~~~~
                              uint32_t
/usr/include/bits/stdint-uintn.h:26:20: note: 'uint32_t' declared here
typedef __uint32_t uint32_t;
                   ^
/tmp/shadow-cast/src/av/sample_format.cpp:36:31: error: no type named 'uint32_t' in namespace 'std'; did you mean simply 'uint32_t'?
auto is_sample_rate_supported(std::uint32_t requested,
                              ^~~~~~~~~~~~~
                              uint32_t
/usr/include/bits/stdint-uintn.h:26:20: note: 'uint32_t' declared here
typedef __uint32_t uint32_t;
                   ^
2 errors generated.
make[2]: *** [src/CMakeFiles/shadow-cast-obj.dir/build.make:174: src/CMakeFiles/shadow-cast-obj.dir/av/sample_format.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:102: src/CMakeFiles/shadow-cast-obj.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

That worked, thanks! Now I'm getting a few other errors related to ffmpeg 6... :sweat_smile: 

```
/tmp/shadow-cast/src/handlers/audio_chunk_writer.cpp:39:12: error: 'channels' is deprecated [-Werror,-Wdeprecated-declarations]
    frame->channels = interleaved ? 2 : chunk.channel_buffers().size();
           ^
/usr/include/libavutil/frame.h:730:5: note: 'channels' has been explicitly marked deprecated here
    attribute_deprecated
    ^
/usr/include/libavutil/attributes.h:100:49: note: expanded from macro 'attribute_deprecated'
#    define attribute_deprecated __attribute__((deprecated))
                                                ^
/tmp/shadow-cast/src/handlers/audio_chunk_writer.cpp:41:12: error: 'channel_layout' is deprecated [-Werror,-Wdeprecated-declarations]
    frame->channel_layout = AV_CH_LAYOUT_STEREO;
           ^
/usr/include/libavutil/frame.h:574:5: note: 'channel_layout' has been explicitly marked deprecated here
    attribute_deprecated
    ^
/usr/include/libavutil/attributes.h:100:49: note: expanded from macro 'attribute_deprecated'
#    define attribute_deprecated __attribute__((deprecated))
                                                ^
2 errors generated.
make[2]: *** [src/CMakeFiles/shadow-cast-obj.dir/build.make:230: src/CMakeFiles/shadow-cast-obj.dir/handlers/audio_chunk_writer.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:102: src/CMakeFiles/shadow-cast-obj.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

I deleted the build directories and now it fails before compiling:

```
$ cmake -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_COMPILER=/usr/bin/clang++
-- The CXX compiler identification is Clang 16.0.6
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "1.8.1") 
-- Pipewire_INCLUDE_DIR: /usr/include/pipewire-0.3
-- Pipewire_LIBRARY: /usr/lib/libpipewire-0.3.so
-- SPA_INCLUDE_DIR: /usr/include/spa-0.2
-- Components: avcodec;avdevice;avfilter;avformat;avutil;swresample;swscale
-- FFMpeg_avcodec: FOUND
-- FFMpeg_avdevice: FOUND
-- FFMpeg_avfilter: FOUND
-- FFMpeg_avformat: FOUND
-- FFMpeg_avutil: FOUND
-- FFMpeg_swresample: FOUND
-- FFMpeg_swscale: FOUND
-- Found X11: /usr/include   
-- Looking for XOpenDisplay in /usr/lib/libX11.so;/usr/lib/libXext.so
-- Looking for XOpenDisplay in /usr/lib/libX11.so;/usr/lib/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Looking for IceConnectionNumber in ICE
-- Looking for IceConnectionNumber in ICE - found
-- Found DRM: /usr/lib/libdrm.so  
-- DRM_INCLUDE_DIRS: /usr/include/libdrm;/usr/include
-- DRM_LIBRARY: /usr/include
-- Change Dir: '/tmp/shadow-cast/build/CMakeFiles/CMakeScratch/TryCompile-futG1P'

Run Build Command(s): /usr/bin/cmake -E env VERBOSE=1 /usr/bin/make -f Makefile cmTC_846a4/fast
/usr/bin/make  -f CMakeFiles/cmTC_846a4.dir/build.make CMakeFiles/cmTC_846a4.dir/build
make[1]: Entering directory '/tmp/shadow-cast/build/CMakeFiles/CMakeScratch/TryCompile-futG1P'
Building CXX object CMakeFiles/cmTC_846a4.dir/ffmpeg_6_check.cpp.o
/usr/bin/clang++   -std=gnu++20 -MD -MT CMakeFiles/cmTC_846a4.dir/ffmpeg_6_check.cpp.o -MF CMakeFiles/cmTC_846a4.dir/ffmpeg_6_check.cpp.o.d -o CMakeFiles/cmTC_846a4.dir/ffmpeg_6_check.cpp.o -c /tmp/shadow-cast/build/CMakeFiles/CMakeScratch/TryCompile-futG1P/ffmpeg_6_check.cpp
In file included from /tmp/shadow-cast/build/CMakeFiles/CMakeScratch/TryCompile-futG1P/ffmpeg_6_check.cpp:2:
In file included from /usr/include/libavutil/avutil.h:301:
/usr/include/libavutil/common.h:30:2: error: missing -D__STDC_CONSTANT_MACROS / #define __STDC_CONSTANT_MACROS
#error missing -D__STDC_CONSTANT_MACROS / #define __STDC_CONSTANT_MACROS
 ^
1 error generated.
make[1]: *** [CMakeFiles/cmTC_846a4.dir/build.make:79: CMakeFiles/cmTC_846a4.dir/ffmpeg_6_check.cpp.o] Error 1
make[1]: Leaving directory '/tmp/shadow-cast/build/CMakeFiles/CMakeScratch/TryCompile-futG1P'
make: *** [Makefile:127: cmTC_846a4/fast] Error 2


-- Configuring done (2.4s)
-- Generating done (0.0s)
-- Build files have been written to: /tmp/shadow-cast/build
```

Not yet:

```
Run Build Command(s): /usr/bin/cmake -E env VERBOSE=1 /usr/bin/make -f Makefile cmTC_1967f/fast
/usr/bin/make  -f CMakeFiles/cmTC_1967f.dir/build.make CMakeFiles/cmTC_1967f.dir/build
make[1]: Entering directory '/tmp/shadow-cast/build/CMakeFiles/CMakeScratch/TryCompile-p1j2hg'
Building CXX object CMakeFiles/cmTC_1967f.dir/ffmpeg_6_check.cpp.o
/usr/bin/clang++   -std=gnu++20 -MD -MT CMakeFiles/cmTC_1967f.dir/ffmpeg_6_check.cpp.o -MF CMakeFiles/cmTC_1967f.dir/ffmpeg_6_check.cpp.o.d -o CMakeFiles/cmTC_1967f.dir/ffmpeg_6_check.cpp.o -c /tmp/shadow-cast/build/CMakeFiles/CMakeScratch/TryCompile-p1j2hg/ffmpeg_6_check.cpp
Linking CXX executable cmTC_1967f
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_1967f.dir/link.txt --verbose=1
/usr/bin/clang++ CMakeFiles/cmTC_1967f.dir/ffmpeg_6_check.cpp.o -o cmTC_1967f 
/usr/bin/ld: CMakeFiles/cmTC_1967f.dir/ffmpeg_6_check.cpp.o: in function `main':
ffmpeg_6_check.cpp:(.text+0x20): undefined reference to `av_channel_layout_default'
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [CMakeFiles/cmTC_1967f.dir/build.make:100: cmTC_1967f] Error 1
make[1]: Leaving directory '/tmp/shadow-cast/build/CMakeFiles/CMakeScratch/TryCompile-p1j2hg'
make: *** [Makefile:127: cmTC_1967f/fast] Error 2
```

Now CMake passes, but building fails:

```
/tmp/shadow-cast/src/handlers/audio_chunk_writer.cpp:44:12: error: 'channels' is deprecated [-Werror,-Wdeprecated-declarations]
    frame->channels = interleaved ? 2 : chunk.channel_buffers().size();
           ^
/usr/include/libavutil/frame.h:730:5: note: 'channels' has been explicitly marked deprecated here
    attribute_deprecated
    ^
/usr/include/libavutil/attributes.h:100:49: note: expanded from macro 'attribute_deprecated'
#    define attribute_deprecated __attribute__((deprecated))
                                                ^
/tmp/shadow-cast/src/handlers/audio_chunk_writer.cpp:45:12: error: 'channel_layout' is deprecated [-Werror,-Wdeprecated-declarations]
    frame->channel_layout = AV_CH_LAYOUT_STEREO;
           ^
/usr/include/libavutil/frame.h:574:5: note: 'channel_layout' has been explicitly marked deprecated here
    attribute_deprecated
    ^
/usr/include/libavutil/attributes.h:100:49: note: expanded from macro 'attribute_deprecated'
#    define attribute_deprecated __attribute__((deprecated))
                                                ^
2 errors generated.
make[2]: *** [src/CMakeFiles/shadow-cast-obj.dir/build.make:230: src/CMakeFiles/shadow-cast-obj.dir/handlers/audio_chunk_writer.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:102: src/CMakeFiles/shadow-cast-obj.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

I think we're really close! After I added these changes:

```diff
diff --git a/src/handlers/audio_chunk_writer.cpp b/src/handlers/audio_chunk_writer.cpp
index 5cd4bcc..db952d2 100644
--- a/src/handlers/audio_chunk_writer.cpp
+++ b/src/handlers/audio_chunk_writer.cpp
@@ -41,7 +41,7 @@ auto ChunkWriter::operator()(MediaChunk const& chunk) -> void
     frame->channels = interleaved ? 2 : chunk.channel_buffers().size();
     frame->channel_layout = AV_CH_LAYOUT_STEREO;
 #else
-    av_channel_layout_copy(&fram->ch_layout, &codec_context->ch_layout);
+    av_channel_layout_copy(&frame->ch_layout, &codec_context_->ch_layout);
 #endif
     frame->pts = total_samples_written_;
     total_samples_written_ += frame->nb_samples;
diff --git a/src/platform/wayland.hpp b/src/platform/wayland.hpp
index 1bc3242..44c6188 100644
--- a/src/platform/wayland.hpp
+++ b/src/platform/wayland.hpp
@@ -2,6 +2,7 @@
 #define SHADOW_CAST_PLATFORM_WAYLAND_HPP_INCLUDED
 
 #include "./egl.hpp"
+#include <cinttypes>
 #include <memory>
 #include <type_traits>
 #include <wayland-client.h>

```

Compilation got to 92%, but it ultimately failed with this error:

```
/tmp/shadow-cast/src/main.cpp:345:37: error: no matching function for call to 'av_buffer_pool_init'
    sc::BufferPoolPtr buffer_pool { av_buffer_pool_init(
                                    ^~~~~~~~~~~~~~~~~~~
/usr/include/libavutil/buffer.h:266:15: note: candidate function not viable: no known conversion from '(lambda at /tmp/shadow-cast/src/main.cpp:346:12)' to 'AVBufferRef *(*)(size_t)' (aka 'AVBufferRef *(*)(unsigned long)') for 2nd argument
AVBufferPool *av_buffer_pool_init(size_t size, AVBufferRef* (*alloc)(size_t size));
              ^
```

--- a/.versioning/changes/a0BeZgsmuY.patch.md
+++ b/.versioning/changes/a0BeZgsmuY.patch.md
@@ -1,0 +1,1 @@
+Fixes build incompatibilities with ffmpeg6 - thankyou, @guihkx (#17)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -35,7 +35,6 @@ find_package(
         avdevice
         avfilter
         avformat
-        avresample
         avutil
         swresample
         swscale

--- a/cmake/FindFFMpeg.cmake
+++ b/cmake/FindFFMpeg.cmake
@@ -85,7 +85,6 @@ try_find_ffmpeg_targets(
         avdevice
         avfilter
         avformat
-        avresample
         avutil
         swresample
         swscale

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -65,7 +65,6 @@ target_link_libraries(shadow-cast-obj
     FFMpeg::avdevice
     FFMpeg::avfilter
     FFMpeg::avformat
-    FFMpeg::avresample
     FFMpeg::avutil
     FFMpeg::swresample
     FFMpeg::swscale

--- a/src/av/codec.cpp
+++ b/src/av/codec.cpp
@@ -21,7 +21,7 @@ auto create_video_encoder(std::string const& encoder_name,
                           std::uint32_t fps,
                           AVPixelFormat pixel_format) -> sc::CodecContextPtr
 {
-    sc::BorrowedPtr<AVCodec> video_encoder { avcodec_find_encoder_by_name(
+    sc::BorrowedPtr<AVCodec const> video_encoder { avcodec_find_encoder_by_name(
         encoder_name.c_str()) };
     if (!video_encoder) {
         throw CodecError { "Failed to find required video codec" };

--- a/src/av/sample_format.hpp
+++ b/src/av/sample_format.hpp
@@ -3,6 +3,7 @@
 
 #include "av/fwd.hpp"
 #include "utils/borrowed_ptr.hpp"
+#include <cinttypes>
 #include <spa/param/audio/format-utils.h>
 #include <span>
 #include <stdexcept>

--- a/src/handlers/audio_chunk_writer.cpp
+++ b/src/handlers/audio_chunk_writer.cpp
@@ -1,6 +1,7 @@
 #include "handlers/audio_chunk_writer.hpp"
 #include "av/frame.hpp"
 #include "av/sample_format.hpp"
+#include "config.hpp"
 #include "error.hpp"
 #include <algorithm>
 #include <cassert>
@@ -36,9 +37,12 @@ auto ChunkWriter::operator()(MediaChunk const& chunk) -> void
     frame->nb_samples = chunk.sample_count;
     frame->format = codec_context_->sample_fmt;
     frame->sample_rate = codec_context_->sample_rate;
+#if LIBAVCODEC_VERSION_MAJOR < 60
     frame->channels = interleaved ? 2 : chunk.channel_buffers().size();
-
     frame->channel_layout = AV_CH_LAYOUT_STEREO;
+#else
+    av_channel_layout_copy(&frame->ch_layout, &codec_context_->ch_layout);
+#endif
     frame->pts = total_samples_written_;
     total_samples_written_ += frame->nb_samples;
 

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -3,6 +3,7 @@
 #include <X11/Xlib.h>
 #include <cassert>
 #include <cmath>
+#include <cstddef>
 #include <cstdio>
 #include <initializer_list>
 #include <iostream>
@@ -10,6 +11,12 @@
 #include <thread>
 #include <type_traits>
 #include <vector>
+
+#if FF_API_BUFFER_SIZE_T
+using BufferSize = int;
+#else
+using BufferSize = std::size_t;
+#endif
 
 template <typename F>
 auto add_signal_handler(sc::Context& ctx, std::uint32_t sig, F&& handler)
@@ -85,7 +92,7 @@ auto run_wayland(sc::Parameters const& params, sc::wayland::DisplayPtr display)
     }
 
     sc::FormatContextPtr format_context { fc_tmp };
-    sc::BorrowedPtr<AVCodec> encoder { avcodec_find_encoder_by_name(
+    sc::BorrowedPtr<AVCodec const> encoder { avcodec_find_encoder_by_name(
         params.audio_encoder.c_str()) };
     if (!encoder) {
         throw sc::CodecError { "Failed to find required audio codec" };
@@ -104,8 +111,12 @@ auto run_wayland(sc::Parameters const& params, sc::wayland::DisplayPtr display)
     if (!audio_encoder_context)
         throw sc::CodecError { "Failed to allocate audio codec context" };
 
+#if LIBAVCODEC_VERSION_MAJOR < 60
     audio_encoder_context->channels = 2;
     audio_encoder_context->channel_layout = av_get_default_channel_layout(2);
+#else
+    av_channel_layout_default(&audio_encoder_context->ch_layout, 2);
+#endif
     audio_encoder_context->sample_rate = params.sample_rate;
     audio_encoder_context->sample_fmt =
         sc::convert_to_libav_format(supported_formats.front());
@@ -282,7 +293,7 @@ auto run(sc::Parameters const& params) -> void
     }
 
     sc::FormatContextPtr format_context { fc_tmp };
-    sc::BorrowedPtr<AVCodec> encoder { avcodec_find_encoder_by_name(
+    sc::BorrowedPtr<AVCodec const> encoder { avcodec_find_encoder_by_name(
         params.audio_encoder.c_str()) };
     if (!encoder) {
         throw sc::CodecError { "Failed to find required audio codec" };
@@ -301,8 +312,12 @@ auto run(sc::Parameters const& params) -> void
     if (!audio_encoder_context)
         throw sc::CodecError { "Failed to allocate audio codec context" };
 
+#if LIBAVCODEC_VERSION_MAJOR < 60
     audio_encoder_context->channels = 2;
     audio_encoder_context->channel_layout = av_get_default_channel_layout(2);
+#else
+    av_channel_layout_default(&audio_encoder_context->ch_layout, 2);
+#endif
     audio_encoder_context->sample_rate = params.sample_rate;
     audio_encoder_context->sample_fmt =
         sc::convert_to_libav_format(supported_formats.front());
@@ -335,7 +350,8 @@ auto run(sc::Parameters const& params) -> void
     }
 
     sc::BufferPoolPtr buffer_pool { av_buffer_pool_init(
-        1, [](int size) { return av_buffer_alloc(size); }) };
+        1, [](BufferSize size) { return av_buffer_alloc(size); }) };
+
     if (!buffer_pool)
         throw sc::CodecError { "Failed to allocate video buffer pool" };
 
@@ -412,7 +428,7 @@ auto run(sc::Parameters const& params) -> void
 
     ctx.services().add_from_factory<sc::VideoService>([&] {
         return std::make_unique<sc::VideoService>(
-            nvcudalib, nvfbc, cuda_ctx.get(), nvfbc_instance.get());
+            nvfbc, cuda_ctx.get(), nvfbc_instance.get());
     });
 
     set_audio_chunk_handler(audio_ctx,

--- a/src/platform/wayland.hpp
+++ b/src/platform/wayland.hpp
@@ -2,6 +2,7 @@
 #define SHADOW_CAST_PLATFORM_WAYLAND_HPP_INCLUDED
 
 #include "./egl.hpp"
+#include <cinttypes>
 #include <memory>
 #include <type_traits>
 #include <wayland-client.h>

--- a/src/services/video_service.cpp
+++ b/src/services/video_service.cpp
@@ -4,12 +4,10 @@ namespace sc
 {
 
 VideoService::VideoService(
-    NvCuda nvcuda,
     NvFBC nvfbc,
     BorrowedPtr<std::remove_pointer_t<CUcontext>> nvcuda_ctx,
     NVFBC_SESSION_HANDLE nvfbc_session) noexcept
-    : nvcuda_ { nvcuda }
-    , nvfbc_ { nvfbc }
+    : nvfbc_ { nvfbc }
     , nvcuda_ctx_ { nvcuda_ctx }
     , nvfbc_session_ { nvfbc_session }
 {
@@ -29,9 +27,7 @@ auto dispatch_frame(Service& svc) -> void
 
     NVFBC_TOCUDA_GRAB_FRAME_PARAMS grab_params {};
     grab_params.dwVersion = NVFBC_TOCUDA_GRAB_FRAME_PARAMS_VER;
-    grab_params.dwFlags =
-        NVFBC_TOCUDA_GRAB_FLAGS_NOWAIT; /* |
-                                           NVFBC_TOCUDA_GRAB_FLAGS_FORCE_REFRESH;*/
+    grab_params.dwFlags = NVFBC_TOCUDA_GRAB_FLAGS_NOWAIT;
     grab_params.pFrameGrabInfo = &frame_info;
     grab_params.pCUDADeviceBuffer = &cu_device_ptr;
     grab_params.dwTimeoutMs = 0;

--- a/src/services/video_service.hpp
+++ b/src/services/video_service.hpp
@@ -16,8 +16,7 @@ struct VideoService final : Service
     using CaptureFrameReceiverType =
         Receiver<void(CUdeviceptr, NVFBC_FRAME_GRAB_INFO)>;
 
-    VideoService(NvCuda,
-                 NvFBC,
+    VideoService(NvFBC,
                  BorrowedPtr<std::remove_pointer_t<CUcontext>>,
                  NVFBC_SESSION_HANDLE) noexcept;
 
@@ -31,7 +30,6 @@ protected:
     auto on_init(ReadinessRegister) -> void override;
 
 private:
-    NvCuda nvcuda_;
     NvFBC nvfbc_;
     BorrowedPtr<std::remove_pointer_t<CUcontext>> nvcuda_ctx_;
     NVFBC_SESSION_HANDLE nvfbc_session_;

